File: mobile/android/native/src/main/java/com/smalltyrant/hocgh/data/UpdateRepository.kt
package com.smalltyrant.hocgh.data

import android.database.sqlite.SQLiteDatabase
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.time.Duration

private const val GITHUB_REPO = "SmallTyrant/hololive_OCG_helper"
private const val DB_RELEASE_TAG = "DB"
private const val DB_RELEASE_API = "https://api.github.com/repos/$GITHUB_REPO/releases/tags/$DB_RELEASE_TAG"
private const val DB_DIRECT_URL = "https://github.com/$GITHUB_REPO/releases/download/$DB_RELEASE_TAG/hololive_ocg.sqlite"
private const val APP_RELEASE_TAG = "install_file1"
private const val APP_APK_ASSET_NAME = "app-release.apk"
private const val APP_APK_DIRECT_URL = "https://github.com/$GITHUB_REPO/releases/download/$APP_RELEASE_TAG/$APP_APK_ASSET_NAME"
private const val APP_VERSION_JSON_URL = "https://github.com/$GITHUB_REPO/releases/download/$APP_RELEASE_TAG/version.json"
private val DB_EXTENSIONS = listOf(".sqlite", ".sqlite3", ".db")

data class ReleaseDbInfo(
    val tag: String,
    val assetName: String,
    val assetUrl: String,
    val assetUpdatedAt: String,
    val assetDigest: String,
    val publishedAt: String,
    val createdAt: String,
) {
    val effectiveDateSource: String
        get() = assetUpdatedAt.ifBlank {
            publishedAt.ifBlank { createdAt }
        }

    val updateMarker: String?
        get() = assetDigest.ifBlank { assetUpdatedAt }.ifBlank { null }
}

data class ReleaseApkInfo(
    val tag: String,
    val assetName: String,
    val assetUrl: String,
    val assetUpdatedAt: String,
    val versionName: String,
    val versionCode: Long,
)

class UpdateRepository {
    private val http = OkHttpClient.Builder()
        .callTimeout(Duration.ofSeconds(120))
        .build()

    fun getLatestReleaseDbInfo(): ReleaseDbInfo {
        val payload = fetchLatestReleasePayload()
        return releaseDbInfoFromPayload(payload)
    }

    fun fetchRemoteDbDate(): String? {
        return runCatching {
            val info = getLatestReleaseDbInfo()
            formatIsoDateOrNull(info.effectiveDateSource)
        }.getOrNull()
    }

    fun fetchLatestApkInfo(): ReleaseApkInfo {
        return runCatching {
            fetchVersionJsonApkInfo()
        }.getOrElse {
            ReleaseApkInfo(
                tag = APP_RELEASE_TAG,
                assetName = APP_APK_ASSET_NAME,
                assetUrl = APP_APK_DIRECT_URL,
                assetUpdatedAt = "",
                versionName = "1.0",
                versionCode = 100,
            )
        }
    }

    fun downloadLatestDb(targetDbFile: File): ReleaseDbInfo {
        val releaseInfo = runCatching { getLatestReleaseDbInfo() }.getOrElse {
            ReleaseDbInfo(
                tag = DB_RELEASE_TAG,
                assetName = "hololive_ocg.sqlite",
                assetUrl = DB_DIRECT_URL,
                assetUpdatedAt = "",
                assetDigest = "",
                publishedAt = "",
                createdAt = "",
            )
        }

        targetDbFile.parentFile?.mkdirs()
        val tempFile = File(targetDbFile.parentFile, "${targetDbFile.name}.download")

        try {
            val request = Request.Builder()
                .url(releaseInfo.assetUrl)
                .header("User-Agent", "hOCG_H/1.1")
                .header("Accept", "application/octet-stream")
                .build()

            http.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw IOException("DB asset HTTP ${response.code}")
                }
                val body = response.body ?: throw IOException("DB asset body is empty")
                tempFile.outputStream().use { output ->
                    body.byteStream().use { input ->
                        input.copyTo(output)
                    }
                }
            }

            validateSqlite(tempFile)
            replaceFile(tempFile, targetDbFile)
            writeReleaseMeta(targetDbFile, releaseInfo)
            return releaseInfo
        } finally {
            if (tempFile.exists()) {
                tempFile.delete()
            }
        }
    }

    private fun fetchLatestReleasePayload(): JSONObject {
        return fetchReleasePayload(DB_RELEASE_API)
    }

    private fun fetchReleasePayload(url: String): JSONObject {
        val request = Request.Builder()
            .url(url)
            .header("User-Agent", "hOCG_H/1.1")
            .header("Accept", "application/vnd.github+json")
            .build()

        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("GitHub API HTTP ${response.code}")
            }
            val body = response.body?.string().orEmpty()
            if (body.isBlank()) {
                throw IOException("GitHub API response body is empty")
            }
            return JSONObject(body)
        }
    }

    private fun releaseDbInfoFromPayload(payload: JSONObject): ReleaseDbInfo {
        val tag = payload.optString("tag_name", "latest")
        val publishedAt = payload.optString("published_at", "")
        val createdAt = payload.optString("created_at", "")
        val assets = payload.optJSONArray("assets") ?: JSONArray()

        val picked = pickAsset(assets)
        val assetName = picked.first
        val assetUrl = picked.second

        var updatedAt = ""
        var digest = ""
        for (i in 0 until assets.length()) {
            val item = assets.optJSONObject(i) ?: continue
            val name = item.optString("name", "")
            val url = item.optString("browser_download_url", "")
            if (name == assetName || url == assetUrl) {
                updatedAt = item.optString("updated_at", "")
                digest = normalizeHash(item.optString("digest", ""))
                break
            }
        }

        return ReleaseDbInfo(
            tag = tag,
            assetName = assetName,
            assetUrl = assetUrl,
            assetUpdatedAt = updatedAt,
            assetDigest = digest,
            publishedAt = publishedAt,
            createdAt = createdAt,
        )
    }

    private fun pickAsset(assets: JSONArray): Pair<String, String> {
        for (preferred in listOf("hololive_ocg.sqlite")) {
            for (i in 0 until assets.length()) {
                val item = assets.optJSONObject(i) ?: continue
                val name = item.optString("name", "")
                val url = item.optString("browser_download_url", "")
                if (name == preferred && url.isNotBlank()) {
                    return name to url
                }
            }
        }

        for (i in 0 until assets.length()) {
            val item = assets.optJSONObject(i) ?: continue
            val name = item.optString("name", "")
            val url = item.optString("browser_download_url", "")
            if (url.isNotBlank() && DB_EXTENSIONS.any { ext -> name.endsWith(ext) }) {
                return name to url
            }
        }

        return "hololive_ocg.sqlite" to DB_DIRECT_URL
    }

    private fun fetchVersionJsonApkInfo(): ReleaseApkInfo {
        val request = Request.Builder()
            .url(APP_VERSION_JSON_URL)
            .header("User-Agent", "hOCG_H/1.1")
            .header("Accept", "application/json")
            .build()

        val payload = http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("Version JSON HTTP ${response.code}")
            }
            val body = response.body?.string().orEmpty()
            if (body.isBlank()) {
                throw IOException("Version JSON body is empty")
            }
            JSONObject(body)
        }

        val versionName = payload.optString("versionName", "").ifBlank {
            payload.optString("version", "1.0")
        }
        val fallbackCode = versionCodeFromName(versionName)
        val versionCode = payload.optLong("versionCode", fallbackCode).takeIf { it > 0 } ?: fallbackCode
        val downloadUrl = payload.optString("apkUrl", "").ifBlank {
            payload.optString("downloadUrl", APP_APK_DIRECT_URL)
        }

        return ReleaseApkInfo(
            tag = payload.optString("tag", APP_RELEASE_TAG),
            assetName = APP_APK_ASSET_NAME,
            assetUrl = downloadUrl,
            assetUpdatedAt = payload.optString("updatedAt", ""),
            versionName = versionName,
            versionCode = versionCode,
        )
    }

    private fun versionCodeFromName(versionName: String): Long {
        val parts = versionName.split('.')
            .mapNotNull { it.toIntOrNull() }
        if (parts.isEmpty()) {
            return 100
        }
        val major = parts.getOrElse(0) { 1 }
        val minor = parts.getOrElse(1) { 0 }
        val patch = parts.getOrElse(2) { 0 }
        return major * 100L + minor * 10L + patch
    }

    private fun normalizeHash(raw: String?): String {
        val value = raw?.trim()?.lowercase().orEmpty()
        return if (value.startsWith("sha256:")) value.removePrefix("sha256:").trim() else value
    }

    private fun validateSqlite(dbFile: File) {
        if (!dbFile.exists() || !dbFile.isFile || dbFile.length() <= 0L) {
            throw IOException("downloaded DB file is missing or empty")
        }

        FileInputStream(dbFile).use { input ->
            val header = ByteArray(16)
            val read = input.read(header)
            if (read != 16 || !header.contentEquals("SQLite format 3\u0000".toByteArray())) {
                throw IOException("downloaded file is not a valid SQLite database")
            }
        }

        SQLiteDatabase.openDatabase(
            dbFile.absolutePath,
            null,
            SQLiteDatabase.OPEN_READONLY,
        ).useDb { db ->
            val hasPrints = db.rawQuery(
                "SELECT 1 FROM sqlite_master WHERE type='table' AND name='prints'",
                null,
            ).useCursor { cursor -> cursor.moveToFirst() }
            if (!hasPrints) {
                throw IOException("downloaded DB is missing prints table")
            }
        }
    }

    private fun writeReleaseMeta(dbFile: File, info: ReleaseDbInfo) {
        runCatching {
            SQLiteDatabase.openDatabase(
                dbFile.absolutePath,
                null,
                SQLiteDatabase.OPEN_READWRITE,
            ).useDb { db ->
                db.execSQL("CREATE TABLE IF NOT EXISTS meta (key TEXT PRIMARY KEY, value TEXT NOT NULL)")

                val values = mapOf(
                    "release_tag" to info.tag,
                    "release_asset_name" to info.assetName,
                    "release_asset_updated_at" to info.assetUpdatedAt,
                    "release_asset_digest" to info.assetDigest,
                    "release_published_at" to info.publishedAt,
                    "release_created_at" to info.createdAt,
                )

                db.beginTransaction()
                try {
                    for ((key, value) in values) {
                        if (value.isBlank()) {
                            continue
                        }
                        db.execSQL(
                            """
                            INSERT INTO meta(key, value)
                            VALUES(?, ?)
                            ON CONFLICT(key) DO UPDATE SET value = excluded.value
                            """.trimIndent(),
                            arrayOf(key, value),
                        )
                    }
                    db.setTransactionSuccessful()
                } finally {
                    db.endTransaction()
                }
            }
        }
    }

    private fun replaceFile(source: File, target: File) {
        if (target.exists() && !target.delete()) {
            throw IOException("failed to replace existing DB file")
        }
        if (!source.renameTo(target)) {
            source.inputStream().use { input ->
                target.outputStream().use { output ->
                    input.copyTo(output)
                }
            }
            source.delete()
        }
    }
}

File: mobile/ios/native/Sources/HocgNative/UpdateRepository.swift
import Foundation
import SQLite3

private let githubRepo = "SmallTyrant/hololive_OCG_helper"
private let latestReleaseAPI = URL(string: "https://api.github.com/repos/\(githubRepo)/releases/latest")!
private let latestDbDirectURL = URL(string: "https://github.com/\(githubRepo)/releases/latest/download/hololive_ocg.sqlite")!

struct ReleaseDbInfo {
    let tag: String
    let assetName: String
    let assetURL: URL
    let assetUpdatedAt: String
    let publishedAt: String
    let createdAt: String
}

final class UpdateRepository {

    func latestReleaseDbInfo() async throws -> ReleaseDbInfo {
        var request = URLRequest(url: latestReleaseAPI)
        request.timeoutInterval = 20
        request.setValue("hOCG_H/1.1", forHTTPHeaderField: "User-Agent")
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }

        let payload = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        guard let payload else {
            throw URLError(.cannotParseResponse)
        }
        return releaseInfo(from: payload)
    }

    func fetchRemoteDbDate() async -> String? {
        guard let info = try? await latestReleaseDbInfo() else {
            return nil
        }
        return formatIsoDateOrNil(
            !info.assetUpdatedAt.isEmpty
                ? info.assetUpdatedAt
                : (!info.publishedAt.isEmpty ? info.publishedAt : info.createdAt)
        )
    }

    func downloadLatestDb(to targetDBURL: URL) async throws -> ReleaseDbInfo {
        let releaseInfo: ReleaseDbInfo
        do {
            releaseInfo = try await latestReleaseDbInfo()
        } catch {
            releaseInfo = ReleaseDbInfo(
                tag: "latest",
                assetName: "hololive_ocg.sqlite",
                assetURL: latestDbDirectURL,
                assetUpdatedAt: "",
                publishedAt: "",
                createdAt: "",
            )
        }

        var request = URLRequest(url: releaseInfo.assetURL)
        request.timeoutInterval = 120
        request.setValue("hOCG_H/1.1", forHTTPHeaderField: "User-Agent")
        request.setValue("application/octet-stream", forHTTPHeaderField: "Accept")

        let (tempDownloadedURL, response) = try await URLSession.shared.download(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }

        let fm = FileManager.default
        let tempTarget = targetDBURL.appendingPathExtension("download")
        if fm.fileExists(atPath: tempTarget.path) {
            try? fm.removeItem(at: tempTarget)
        }
        if fm.fileExists(atPath: targetDBURL.deletingLastPathComponent().path) == false {
            try fm.createDirectory(at: targetDBURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        }

        try fm.copyItem(at: tempDownloadedURL, to: tempTarget)

        do {
            try validateSQLite(fileURL: tempTarget)
            if fm.fileExists(atPath: targetDBURL.path) {
                try fm.removeItem(at: targetDBURL)
            }
            try fm.moveItem(at: tempTarget, to: targetDBURL)
            try writeReleaseMeta(dbURL: targetDBURL, info: releaseInfo)
            return releaseInfo
        } catch {
            try? fm.removeItem(at: tempTarget)
            throw error
        }
    }

    private func releaseInfo(from payload: [String: Any]) -> ReleaseDbInfo {
        let tag = payload["tag_name"] as? String ?? "latest"
        let publishedAt = payload["published_at"] as? String ?? ""
        let createdAt = payload["created_at"] as? String ?? ""

        let assets = payload["assets"] as? [[String: Any]] ?? []
        let asset = pickAsset(from: assets)

        let assetName = asset.name
        let assetURL = asset.url

        var assetUpdatedAt = ""
        for item in assets {
            let name = item["name"] as? String ?? ""
            let urlString = item["browser_download_url"] as? String ?? ""
            if (name == assetName || urlString == assetURL.absoluteString),
               let updated = item["updated_at"] as? String {
                assetUpdatedAt = updated
                break
            }
        }

        return ReleaseDbInfo(
            tag: tag,
            assetName: assetName,
            assetURL: assetURL,
            assetUpdatedAt: assetUpdatedAt,
            publishedAt: publishedAt,
            createdAt: createdAt,
        )
    }

    private func pickAsset(from assets: [[String: Any]]) -> (name: String, url: URL) {
        for item in assets {
            let name = item["name"] as? String ?? ""
            let urlString = item["browser_download_url"] as? String ?? ""
            if name == "hololive_ocg.sqlite", let url = URL(string: urlString) {
                return (name, url)
            }
        }

        for item in assets {
            let name = item["name"] as? String ?? ""
            let urlString = item["browser_download_url"] as? String ?? ""
            if [".sqlite", ".sqlite3", ".db"].contains(where: { name.hasSuffix($0) }),
               let url = URL(string: urlString) {
                return (name, url)
            }
        }

        return ("hololive_ocg.sqlite", latestDbDirectURL)
    }

    private func validateSQLite(fileURL: URL) throws {
        let data = try Data(contentsOf: fileURL, options: .mappedIfSafe)
        guard data.count > 16 else {
            throw URLError(.cannotDecodeContentData)
        }
        let header = Data("SQLite format 3\u{0}".utf8)
        guard data.prefix(16) == header else {
            throw URLError(.cannotDecodeContentData)
        }

        let hasPrints = try withSQLite(path: fileURL.path, readOnly: true) { db in
            let stmt = try sqlitePrepare(
                db: db,
                sql: "SELECT 1 FROM sqlite_master WHERE type='table' AND name='prints'",
            )
            defer { sqlite3_finalize(stmt) }
            return sqlite3_step(stmt) == SQLITE_ROW
        }

        if !hasPrints {
            throw URLError(.cannotDecodeContentData)
        }
    }

    private func writeReleaseMeta(dbURL: URL, info: ReleaseDbInfo) throws {
        try withSQLite(path: dbURL.path, readOnly: false) { db in
            if sqlite3_exec(
                db,
                "CREATE TABLE IF NOT EXISTS meta (key TEXT PRIMARY KEY, value TEXT NOT NULL)",
                nil,
                nil,
                nil
            ) != SQLITE_OK {
                throw SQLiteError.execute(sqliteErrorMessage(db))
            }

            let values: [String: String] = [
                "release_tag": info.tag,
                "release_asset_name": info.assetName,
                "release_asset_updated_at": info.assetUpdatedAt,
                "release_published_at": info.publishedAt,
                "release_created_at": info.createdAt,
            ]

            sqlite3_exec(db, "BEGIN TRANSACTION", nil, nil, nil)
            do {
                for (key, value) in values where !value.isEmpty {
                    let stmt = try sqlitePrepare(
                        db: db,
                        sql: """
                        INSERT INTO meta(key, value)
                        VALUES(?, ?)
                        ON CONFLICT(key) DO UPDATE SET value = excluded.value
                        """,
                    )
                    defer { sqlite3_finalize(stmt) }
                    try sqliteBind([.text(key), .text(value)], to: stmt)
                    if sqlite3_step(stmt) != SQLITE_DONE {
                        throw SQLiteError.step(sqliteErrorMessage(db))
                    }
                }
                sqlite3_exec(db, "COMMIT", nil, nil, nil)
            } catch {
                sqlite3_exec(db, "ROLLBACK", nil, nil, nil)
                throw error
            }
        }
    }
}
